### PR TITLE
[Feature] support iceberg partitioned table sink

### PR DESCRIPTION
**Pull request by stephen-shelby**
## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes:

Fixes #
https://github.com/StarRocks/starrocks/issues/21502
## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: be/src/exec/pipeline/pipeline_builder.cpp [hunk 1]
```diff
@@ -153,6 +153,36 @@ void PipelineBuilderContext::maybe_interpolate_local_passthrough_exchange_for_si
...
+                                                                                 pseudo_plan_node_id, local_exchanger);
+
+    _fragment_context->pipelines().back()->add_op_factory(local_shuffle_sink);
+
+    OpFactories operators_source_with_local_shuffle;
+    local_shuffle_source->set_runtime_state(state);
+    local_shuffle_source->set_group_leader(source_operator);
+    local_shuffle_source->set_degree_of_parallelism(desired_sink_dop);
+    operators_source_with_local_shuffle.emplace_back(std::move(local_shuffle_source));
+    operators_source_with_local_shuffle.emplace_back(std::move(table_sink_operator));
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/pipeline_builder.cpp:179]: Could simplify the creation of `operators_source_with_local_shuffle`.
```suggestion
    local_shuffle_source->set_runtime_state(state);
    local_shuffle_source->set_group_leader(source_operator);
    local_shuffle_source->set_degree_of_parallelism(desired_sink_dop);
    OpFactories operators_source_with_local_shuffle{std::move(local_shuffle_source),  std::move(table_sink_operator)};
```

Comment (ZiheLiu) [inline be/src/exec/pipeline/pipeline_builder.cpp:179]: Could simplify the creation of `operators_source_with_local_shuffle`.
```suggestion
    local_shuffle_source->set_runtime_state(state);
    local_shuffle_source->set_group_leader(source_operator);
    local_shuffle_source->set_degree_of_parallelism(desired_sink_dop);
    OpFactories operators_source_with_local_shuffle{std::move(local_shuffle_source),  std::move(table_sink_operator)};
```

File: be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp [hunk 2]
```diff
@@ -92,62 +93,55 @@ Status IcebergTableSinkOperator::push_chunk(RuntimeState* state, const ChunkPtr&
...
-        //        _partition_writers[partition_location]->append_chunk(chunk.get(), state);
-        //
+        Columns partitions_columns;
+        partitions_columns.resize(_partition_expr.size());
+        for (size_t i = 0; i < partitions_columns.size(); ++i) {
+            ASSIGN_OR_RETURN(partitions_columns[i], _partition_expr[i]->evaluate(chunk.get()));
+            DCHECK(partitions_columns[i] != nullptr);
+        }
+
+        std::vector<std::string> partition_column_names = _iceberg_table->partition_column_names();
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp:112]: Could be `const std::vector<std::string>&`?

File: be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp [hunk 3]
```diff
@@ -92,62 +93,55 @@ Status IcebergTableSinkOperator::push_chunk(RuntimeState* state, const ChunkPtr&
...
+        }
+
+        DCHECK(partition_column_names.size() == partition_column_values.size());
+
+        string partition_location = _get_partition_location(partition_column_names, partition_column_values);
+
+        auto partition_writer = _partition_writers.find(partition_location);
+        if (partition_writer == _partition_writers.end()) {
+            tableInfo.partition_location = partition_location;
+            auto writer = new RollingAsyncParquetWriter(tableInfo, _output_expr, _common_metrics.get(),
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp:125]: Who frees this `writer`?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 4 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/InsertPlanner.java|0|4|00.00%|[670, 671, 672, 674]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 4 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/InsertPlanner.java|0|4|00.00%|[670, 671, 672, 674]|

**Review by ZiheLiu — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Pull request merged**